Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs

```diff
@@ -121,9 +142,162 @@ tfFulfillments =
+tfPayments booking =
+  Just
+    [ Spec.Payment
+        { paymentCollectedBy = Just $ show Enums.BPP,
+          paymentId = Nothing,
+          paymentParams = mkParams,
+          paymentStatus = Nothing,
+          paymentTags = Nothing,
+          paymentType = Just $ show Enums.ON_FULFILLMENT
+        }
+    ]
+  where
+    mkParams =
+      Just
+        Spec.PaymentParams
+          { paymentParamsAmount = Just $ encodeToText booking.estimatedFare,
+            paymentParamsBankAccountNumber = Nothing,
+            paymentParamsBankCode = Nothing,
+            paymentParamsCurrency = Just "INR",
+            paymentParamsVirtualPaymentAddress = Nothing
+          }
+
+tfItems :: (CacheFlow m r, EsqDBFlow m r) => DRB.Booking -> m (Maybe [Spec.Item])
+tfItems booking = do
+  merchant <- CQM.findById booking.providerId >>= fromMaybeM (MerchantNotFound booking.providerId.getId)
+  pure $
+    Just
+      [ Spec.Item
+          { itemDescriptor = Nothing,
+            itemFulfillmentIds = Just [booking.quoteId],
+            itemId = Just $ Common.mkItemId merchant.shortId.getShortId booking.vehicleVariant,
+            itemLocationIds = Nothing,
+            itemPaymentIds = Nothing,
+            itemPrice = tfItemPrice booking,
+            itemTags = Nothing
+          }
+      ]
+
+tfItemPrice :: DRB.Booking -> Maybe Spec.Price
+tfItemPrice booking =
+  Just
+    Spec.Price
+      { priceComputedValue = Nothing,
+        priceCurrency = Just "INR",
+        priceMaximumValue = Nothing,
+        priceMinimumValue = Nothing,
+        priceOfferedValue = Nothing,
+        priceValue = Just $ encodeToText booking.estimatedFare
+      }
+
+tfVehicle :: DRB.Booking -> Maybe Spec.Vehicle
+tfVehicle booking = do
+  let (category, variant) = BUtils.castVariant booking.vehicleVariant
+  Just $
+    Spec.Vehicle
+      { vehicleCategory = Just category,
+        vehicleVariant = Just variant,
+        vehicleColor = Nothing,
+        vehicleMake = Nothing,
+        vehicleModel = Nothing,
+        vehicleRegistration = Nothing
+      }
+
+tfAgent :: (CacheFlow m r, EsqDBFlow m r) => DRB.Booking -> m (Maybe Spec.Agent)
+tfAgent booking = do
+  driverQuote <- QDQ.findById (Id booking.quoteId) >>= fromMaybeM (QuoteNotFound booking.quoteId)
+  pure $
+    Just $
+      Spec.Agent
+        { agentContact = Nothing,
+          agentPerson =
+            Just
+              Spec.Person
+                { personId = Nothing,
+                  personImage = Nothing,
+                  personName = Just driverQuote.driverName,
+                  personTags = Nothing
+                }
+        }
+
+tfCustomer :: (CacheFlow m r, EsqDBFlow m r, EncFlow m r) => DRB.Booking -> m (Maybe Spec.Customer)
+tfCustomer booking = do
+  riderId <- booking.riderId & fromMaybeM (BookingFieldNotPresent "rider_id")
+  riderDetails <- runInReplica $ QRiderDetails.findById riderId >>= fromMaybeM (RiderDetailsNotFound riderId.getId)
```

Comment:
Don't make DB calls in transformer functions. Keep the transformer functions pure.